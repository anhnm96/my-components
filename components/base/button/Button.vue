<script lang="ts" setup>
import Spinner from '../Spinner.vue'

const props = withDefaults(
  defineProps<{
    loading?: boolean
    loadingMsg?: string
    success?: boolean
    successMsg?: string
  }>(),
  {
    loading: false,
    loadingMsg: 'processing, wait...',
    success: false,
  }
)
const emit = defineEmits<{
  (e: 'click', ev: MouseEvent): void
}>()

const btnRef = ref<HTMLButtonElement>()
function click(event: MouseEvent) {
  const isBtnDisabled = btnRef.value?.getAttribute('aria-disabled') === 'true'
  if (isBtnDisabled || props.loading) {
    return
  }
  emit('click', event)
}
</script>

<template>
  <button
    ref="btnRef"
    class="btn"
    :class="[(loading || success) && '!pointer-events-none !text-transparent']"
    @click="click"
  >
    <slot />
    <Transition name="fade" mode="out-in">
      <div
        v-if="loading"
        class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
      >
        <span v-if="loadingMsg" class="sr-only" aria-live="assertive">
          {{ loadingMsg }}
        </span>
        <Spinner />
      </div>
      <div
        v-else-if="success"
        class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
      >
        <span v-if="successMsg" class="sr-only" aria-live="assertive">
          {{ successMsg }}
        </span>
        <Icon name="ic:baseline-check" class="text-white" />
      </div>
      <div v-else></div>
    </Transition>
  </button>
</template>

<style scoped>
button {
  @apply relative;
}

button:disabled,
button[aria-disabled='true'] {
  @apply cursor-not-allowed opacity-70;
}
</style>
