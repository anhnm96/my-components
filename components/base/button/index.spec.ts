import { shallowMount } from '@vue/test-utils'
import Button from './index.vue'

let wrapper: any
describe('Button.vue', () => {
  beforeEach(() => {
    wrapper = shallowMount(Button)
  })

  it('is called', () => {
    expect(wrapper.exists()).toBeTruthy()
  })
})
