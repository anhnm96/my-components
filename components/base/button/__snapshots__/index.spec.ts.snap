// Vitest Snapshot v1

exports[`Button.vue > render correctly 1`] = `
"<button class=\\"btn\\" data-v-cbcdd221=\\"\\">
  <!--v-if-->
</button>"
`;
